File: src/tokenizer.py
from utils.enums.tokens import Tokens
from utils.enums.types import Types
from utils.token import Token


class Tokenizer:
    def __init__(self):
        self.source = ""
        self.start = [0, 0]
        self.current = [0, 0]
        self.line = 1
        self.tokens = []
        self.keywords = {
            "and": Tokens.AND,
            "or": Tokens.OR,
            "xor": Tokens.XOR,
            "not": Tokens.NOT,
            "lshift": Tokens.LEFT_SHIFT,
            "rshift": Tokens.RIGHT_SHIFT,
            "if": Tokens.IF,
            "else": Tokens.ELSE,
            "while": Tokens.WHILE,
            "for": Tokens.FOR,
            "break": Tokens.BREAK,
            "continue": Tokens.CONTINUE,
            "match": Tokens.MATCH,
            "case": Tokens.CASE,
            "unwrap": Tokens.UNWRAP,
            "handle": Tokens.HANDLE,
            "function": Tokens.FUNCTION,
            "return": Tokens.RETURN,
            "call": Tokens.CALL,
            "do": Tokens.DO,
            "with": Tokens.WITH,
            "const": Tokens.CONST,
            "static": Tokens.STATIC,
            "mutable": Tokens.MUTABLE,
            "set": Tokens.SET,
            "struct": Tokens.STRUCT,
            "implement": Tokens.IMPLEMENT,
            "enum": Tokens.ENUM,
            "union": Tokens.UNION,
            "pop": Tokens.POP,
            "drop": Tokens.DROP,
            "dup": Tokens.DUP,
            "over": Tokens.OVER,
            "rot": Tokens.ROT,
            "swap": Tokens.SWAP,
            "require": Tokens.REQUIRE,
            "defer": Tokens.DEFER,
            "end": Tokens.END,
            "true": Tokens.BOOLEAN,
            "false": Tokens.BOOLEAN,
        }
        self.keywords.update({
            type_kind.name.lower(): Tokens.TYPE for type_kind in Types
        })

    def tokenize(self, source):
        self.source = source

        while not self.__eof():
            self.start[0] = self.current[0]
            self.start[1] = self.current[1]
            self.__tokenize_lexeme()

        return self.tokens

    def __tokenize_lexeme(self):
        lexeme = self.__advance()

        match lexeme:
            case " " | "\t":
                pass

            case "\n":
                self.line += 1
                self.current[1] = 0

            case "#":
                while not self.__eof() and self.__peek() != "\n":
                    self.__advance()

            case "(":
                self.__add_token(Tokens.LEFT_PAREN)
            case ")":
                self.__add_token(Tokens.RIGHT_PAREN)
            case "{":
                self.__add_token(Tokens.LEFT_CURLY)
            case "}":
                self.__add_token(Tokens.RIGHT_CURLY)
            case "[":
                self.__add_token(Tokens.LEFT_SQUARE)
            case "]":
                self.__add_token(Tokens.RIGHT_SQUARE)
            case ":":
                self.__add_token(Tokens.COLON)
            case ";":
                self.__add_token(Tokens.SEMI_COLON)
            case ",":
                self.__add_token(Tokens.COMMA)
            case ".":
                self.__add_token(Tokens.DOT)
            case "?":
                self.__add_token(Tokens.QUESTION)
            case "%":
                self.__add_token(Tokens.PERCENT)
            case "&":
                self.__add_token(Tokens.AMPERSAND)
            case "|":
                self.__add_token(Tokens.BAR)
            case "*":
                self.__add_token(Tokens.ASTERISK)
            case "^":
                self.__add_token(Tokens.CARET)

            case "/":
                self.__add_token(
                    Tokens.SLASH_SLASH if self.__match("/") else Tokens.SLASH
                )
            case "=":
                self.__add_token(
                    Tokens.EQUAL_EQUAL if self.__match("=") else Tokens.EQUAL
                )
            case "<":
                self.__add_token(
                    Tokens.LESS_EQUAL if self.__match("=") else Tokens.LESS
                )
            case ">":
                self.__add_token(
                    Tokens.GREATER_EQUAL if self.__match("=") else Tokens.GREATER
                )
            case "!":
                self.__add_token(
                    Tokens.NOT_EQUAL if self.__match("=") else Tokens.EXCLAMATION
                )

            case '"':
                self.__handle_strings()

            case "'":
                self.__handle_runes()

            case "-":
                if not self.__eof() and self.__peek().isdigit():
                    self.__handle_numbers()
                else:
                    self.__add_token(Tokens.MINUS)

            case "+":
                if not self.__eof() and self.__peek().isdigit():
                    self.__handle_numbers()
                else:
                    self.__add_token(Tokens.PLUS)

            case _ if lexeme.isdigit():
                self.__handle_numbers()

            case _ if lexeme.isalpha() or lexeme in ["_", "@"]:
                self.__handle_identifiers()

            case _:
                # FIXME: add warning
                pass

    def __handle_numbers(self):
        while not self.__eof() and (self.__peek().isdigit() or self.__peek() in ["_", ","]):
            self.__advance()

        if (
            not self.__eof()
            and self.__peek() == "."
            and (self.__peek_next().isdigit() or self.__peek_next() in ["_", ","])
        ):
            self.__advance()

            while not self.__eof() and (self.__peek().isdigit() or self.__peek() in ["_", ","]):
                self.__advance()

            self.__add_token(Tokens.FLOAT)
        else:
            self.__add_token(Tokens.INTEGER)

    def __handle_strings(self):
        while not self.__eof() and self.__peek() != '"':
            if self.__peek() == "\n":
                # FIXME: add error
                pass

            if self.__match("\\"):
                if self.__eof():
                    # FIXME: add error
                    pass

                escape_rune = self.__peek()
                if escape_rune in ["\\", "'", '"', "n", "r", "t", "v", "b", "a", "f"]:
                    self.__advance()
                else:
                    # FIXME: add error
                    pass
            else:
                self.__advance()

        if self.__eof() or not self.__match('"'):
            # FIXME: add error
            pass

        self.__add_token(Tokens.STRING)

    def __handle_runes(self):
        while not self.__eof() and self.__peek() != "'":
            if self.__peek() == "\n":
                # FIXME: add error
                pass

            if self.__peek() == "\\":
                self.__advance()

                if self.__eof():
                    # FIXME: add error
                    pass

                escape_rune = self.__peek()
                if escape_rune in ["\\", "'", '"', "n", "r", "t", "v", "b", "a", "f"]:
                    self.__advance()
                else:
                    # FIXME: add error
                    pass
            else:
                self.__advance()

        if self.__eof() or not self.__match("'"):
            # FIXME: add error
            pass

        content = self.source[self.start + 1 : self.current - 1]
        if len(content.replace("\\", "")) > 1:
            # FIXME: add error
            pass

        self.__add_token(Tokens.RUNE)

    def __handle_identifiers(self):
        while not self.__eof() and (self.__peek().isalnum() or self.__peek() == "_"):
            self.__advance()

        text = self.source[self.start[0] : self.current[0]]
        token_type = self.keywords.get(text.lower()) or Tokens.IDENTIFIER
        self.__add_token(token_type)

    def __add_token(self, token_kind):
        self.tokens.append(
            Token(
                token_kind,
                self.source[self.start[0] : self.current[0]],
                self.__get_location(),
            )
        )

    def __advance(self):
        lexeme = self.__peek()
        self.current[0] += 1
        self.current[1] += 1
        return lexeme

    def __match(self, expected):
        if self.__eof() or self.__peek() != expected:
            return False
        self.__advance()
        return True

    def __get_location(self):
        return (self.line, self.start[1], self.current[1])

    def __eof(self):
        return self.current[0] >= len(self.source)

    def __peek(self):
        return self.source[self.current[0]]

    def __peek_next(self):
        return self.source[self.current[0] + 1]

File: core/tokenizer.lua
local Tokens = require("utils.enums.tokens")
local Types = require("utils.enums.types")
local Token = require("utils.token")


local Tokenizer = {}
Tokenizer.__index = Tokenizer

function Tokenizer.new(args)
    local log     = args[1] or args.log

    local self    = setmetatable({}, Tokenizer)
    self.log      = log
    self.source   = ""
    self.start    = { 0, 0 }
    self.current  = { 0, 0 }
    self.line     = 1
    self.tokens   = {}
    self.keywords = {
        ["and"] = Tokens.AND,
        ["or"] = Tokens.OR,
        ["xor"] = Tokens.XOR,
        ["not"] = Tokens.NOT,
        ["lshift"] = Tokens.LEFT_SHIFT,
        ["rshift"] = Tokens.RIGHT_SHIFT,
        ["if"] = Tokens.IF,
        ["else"] = Tokens.ELSE,
        ["while"] = Tokens.WHILE,
        ["for"] = Tokens.FOR,
        ["break"] = Tokens.BREAK,
        ["continue"] = Tokens.CONTINUE,
        ["match"] = Tokens.MATCH,
        ["case"] = Tokens.CASE,
        ["unwrap"] = Tokens.UNWRAP,
        ["handle"] = Tokens.HANDLE,
        ["function"] = Tokens.FUNCTION,
        ["return"] = Tokens.RETURN,
        ["call"] = Tokens.CALL,
        ["do"] = Tokens.DO,
        ["with"] = Tokens.WITH,
        ["const"] = Tokens.CONST,
        ["static"] = Tokens.STATIC,
        ["mutable"] = Tokens.MUTABLE,
        ["set"] = Tokens.SET,
        ["struct"] = Tokens.STRUCT,
        ["implement"] = Tokens.IMPLEMENT,
        ["enum"] = Tokens.ENUM,
        ["union"] = Tokens.UNION,
        ["pop"] = Tokens.POP,
        ["drop"] = Tokens.DROP,
        ["dup"] = Tokens.DUP,
        ["over"] = Tokens.OVER,
        ["rot"] = Tokens.ROT,
        ["swap"] = Tokens.SWAP,
        ["require"] = Tokens.REQUIRE,
        ["defer"] = Tokens.DEFER,
        ["end"] = Tokens.END,
        ["true"] = Tokens.BOOLEAN,
        ["false"] = Tokens.BOOLEAN,
    }

    for name, _ in pairs(Types) do
        self.keywords[name:lower()] = Tokens.TYPE
    end

    return self
end

function Tokenizer:tokenize(args)
    local source = args[1] or args.source

    self.source = source

    while not self:_eof() do
        self.start[1] = self.current[1]
        self.start[2] = self.current[2]

        local lexeme = self:_advance()

        if lexeme == " " or lexeme == "\t" then
        elseif lexeme == "\n" then
            self.line = self.line + 1
            self.current[2] = 0
        elseif lexeme == "#" then
            while not self:_eof() and self:_peek() ~= "\n" do
                self:_advance()
            end
        elseif lexeme == "(" then
            self:_add_token(Tokens.LEFT_PAREN)
        elseif lexeme == ")" then
            self:_add_token(Tokens.RIGHT_PAREN)
        elseif lexeme == "{" then
            self:_add_token(Tokens.LEFT_CURLY)
        elseif lexeme == "}" then
            self:_add_token(Tokens.RIGHT_CURLY)
        elseif lexeme == "[" then
            self:_add_token(Tokens.LEFT_SQUARE)
        elseif lexeme == "]" then
            self:_add_token(Tokens.RIGHT_SQUARE)
        elseif lexeme == ":" then
            self:_add_token(Tokens.COLON)
        elseif lexeme == ";" then
            self:_add_token(Tokens.SEMI_COLON)
        elseif lexeme == "," then
            self:_add_token(Tokens.COMMA)
        elseif lexeme == "." then
            self:_add_token(Tokens.DOT)
        elseif lexeme == "?" then
            self:_add_token(Tokens.QUESTION)
        elseif lexeme == "%" then
            self:_add_token(Tokens.PERCENT)
        elseif lexeme == "&" then
            self:_add_token(Tokens.AMPERSAND)
        elseif lexeme == "|" then
            self:_add_token(Tokens.BAR)
        elseif lexeme == "*" then
            self:_add_token(Tokens.ASTERISK)
        elseif lexeme == "^" then
            self:_add_token(Tokens.CARET)
        elseif lexeme == "/" then
            if self:_match("/") then
                self:_add_token(Tokens.SLASH_SLASH)
            else
                self:_add_token(Tokens.SLASH)
            end
        elseif lexeme == "=" then
            if self:_match("=") then
                self:_add_token(Tokens.EQUAL_EQUAL)
            else
                self:_add_token(Tokens.EQUAL)
            end
        elseif lexeme == "<" then
            if self:_match("=") then
                self:_add_token(Tokens.LESS_EQUAL)
            else
                self:_add_token(Tokens.LESS)
            end
        elseif lexeme == ">" then
            if self:_match("=") then
                self:_add_token(Tokens.GREATER_EQUAL)
            else
                self:_add_token(Tokens.GREATER)
            end
        elseif lexeme == "!" then
            if self:_match("=") then
                self:_add_token(Tokens.NOT_EQUAL)
            else
                self:_add_token(Tokens.EXCLAMATION)
            end
        elseif lexeme == '"' then
            self:_handle_strings()
        elseif lexeme == "'" then
            self:_handle_runes()
        elseif lexeme == "-" or lexeme == "+" then
            if not self:_eof() and self:_peek():match("%d") then
                self:_handle_numbers()
            else
                self:_add_token(lexeme == "-" and Tokens.MINUS or Tokens.PLUS)
            end
        elseif lexeme:match("%d") then
            self:_handle_numbers()
        elseif lexeme:match("[a-zA-Z]") or lexeme == "_" or lexeme == "@" then
            self:_handle_identifiers()
        else
            self.log:print({
                "warning",
                "Unsupported symbol",
                location = self:_get_location(),
                code = "W001"
            })
        end
    end

    return self.tokens
end

function Tokenizer:_handle_numbers()
    while not self:_eof() and (self:_peek():match("[0-9_,]")) do
        self:_advance()
    end

    if not self:_eof() and self:_peek() == "." and self:_peek_next():match("[0-9_,]") then
        self:_advance()
        while not self:_eof() and (self:_peek():match("[0-9_,]")) do
            self:_advance()
        end
        self:_add_token(Tokens.FLOAT)
    else
        self:_add_token(Tokens.INTEGER)
    end
end

function Tokenizer:_handle_strings()
    while not self:_eof() and self:_peek() ~= '"' do
        if self:_peek() == "\n" then
            self.log:print({
                "error",
                "Invalid string syntax",
                location = self:_get_location(),
                information = "new lines are not supported inside strings",
                code = "E120"
            })
            os.exit(120)
        end

        if self:_match("\\") then
            if self:_eof() then
                self.log:print({
                    "error",
                    "Invalid string syntax",
                    location = { self.line, self.current[2] - 1, self.current[2] },
                    information = "escape symbols should be followed by a valid letter",
                    code = "E121"
                })
                os.exit(121)
            end

            local escape_rune = self:_peek()
            if escape_rune:match("[\\\"'nrtvabf]") then
                self:_advance()
            else
                self.log:print({
                    "error",
                    "Invalid string syntax",
                    location = { self.line, self.current[2] - 1, self.current[2] + 1 },
                    information = "escape symbols should be followed by a valid letter",
                    code = "E121"
                })
                os.exit(121)
            end
        else
            self:_advance()
        end
    end

    if self:_eof() or not self:_match('"') then
        self.log:print({
            "error",
            "Invalid string syntax",
            location = self:_get_location(),
            information = "string literals need to be closed with a corresponding quote",
            code = "E122"
        })
        os.exit(122)
    end

    self:_add_token(Tokens.STRING)
end

function Tokenizer:_handle_runes()
    while not self:_eof() and self:_peek() ~= "'" do
        if self:_peek() == "\n" then
            self.log:print({
                "error",
                "Invalid rune syntax",
                location = self:_get_location(),
                information = "new lines are not supported inside runes",
                code = "E130"
            })
            os.exit(130)
        end

        if self:_peek() == "\\" then
            self:_advance()
            if self:_eof() then
                self.log:print({
                    "error",
                    "Invalid rune syntax",
                    location = { self.line, self.current[2] - 1, self.current[2] },
                    information = "escape symbols should be followed by a valid letter",
                    code = "E131"
                })
                os.exit(131)
            end

            local escape_rune = self:_peek()
            if escape_rune:match("[\\\'\"nrtvabf]") then
                self:_advance()
            else
                self.log:print({
                    "error",
                    "Invalid rune syntax",
                    location = { self.line, self.current[2] - 1, self.current[2] + 1 },
                    information = "escape symbols should be followed by a valid letter",
                    code = "E131"
                })
                os.exit(131)
            end
        else
            self:_advance()
        end
    end

    if self:_eof() or not self:_match("'") then
        self.log:print({
            "error",
            "Invalid rune syntax",
            location = self:_get_location(),
            information = "rune literals need to be closed with a corresponding quote",
            code = "E132"
        })
        os.exit(132)
    end

    local content = self.source:sub(self.start[1] + 2, self.current[1] - 1)
    if #content:gsub("\\.", "") > 1 then
        self.log:print({
            "error",
            "Invalid rune syntax",
            location = self:_get_location(),
            information = "runes should only contain a letter or an escape sequence",
            code = "E133"
        })
        os.exit(133)
    end

    self:_add_token(Tokens.RUNE)
end

function Tokenizer:_handle_identifiers()
    while not self:_eof() and (self:_peek():match("[a-zA-Z0-9_@]")) do
        self:_advance()
    end

    local text = self.source:sub(self.start[1] + 1, self.current[1])
    local token_type = self.keywords[text:lower()] or Tokens.IDENTIFIER
    self:_add_token(token_type)
end

function Tokenizer:_add_token(token_kind)
    table.insert(self.tokens, Token.new(
        token_kind,
        self.source:sub(self.start[1] + 1, self.current[1]),
        self:_get_location()
    ))
end

function Tokenizer:_advance()
    local lexeme = self:_peek()
    self.current[1] = self.current[1] + 1
    self.current[2] = self.current[2] + 1
    return lexeme
end

function Tokenizer:_match(expected)
    if self:_eof() or self:_peek() ~= expected then
        return false
    end
    self:_advance()
    return true
end

function Tokenizer:_get_location()
    return { self.line, self.start[2], self.current[2] }
end

function Tokenizer:_eof()
    return self.current[1] >= #self.source
end

function Tokenizer:_peek()
    return self.source:sub(self.current[1] + 1, self.current[1] + 1)
end

function Tokenizer:_peek_next()
    return self.source:sub(self.current[1] + 2, self.current[1] + 2)
end

return Tokenizer
